### ISSUE TITLE
Rich results returning errors for products with price range

### ISSUE THREAD

**Issue by sacr3dc0w** [bigcommerce/cornerstone#2153]
### Expected behavior
Structured data to be parsed by Google without error.

### Actual behavior
If conditional of `{{#if product.price.price_range}}` is met, an error gets thrown, as there's no price for the Offer.

### Steps to reproduce behavior
Always displaying price does the trick for me.

Link to rich results (I'll leave up for 24 hours, but will be pushing the fix below definitely before Wednesday): https://search.google.com/test/rich-results/result?id=f09f3CeP4GFgAlRExDpa9w

templates/components/products/schema.html

**BEFORE:**
```
        "offers": {
            "@type": "Offer",
            "priceCurrency": "{{currency_selector.active_currency_code}}",
            {{#if product.price.price_range}}
            "minPrice": "{{#if product.price.price_range.min.with_tax}}{{product.price.price_range.min.with_tax.value}}{{else}}{{product.price.price_range.min.without_tax.value}}{{/if}}",
            "maxPrice": "{{#if product.price.price_range.max.with_tax}}{{product.price.price_range.max.with_tax.value}}{{else}}{{product.price.price_range.max.without_tax.value}}{{/if}}",
            {{else}}
            "price": "{{#if product.price.with_tax}}{{product.price.with_tax.value}}{{else}}{{product.price.without_tax.value}}{{/if}}",
            {{/if}}
            "itemCondition" : "https://schema.org/{{#if product.condition}}{{product.condition}}{{else}}New{{/if}}Condition",
            "availability" : "https://schema.org/{{#if product.pre_order}}PreOrder{{else if product.out_of_stock}}OutOfStock{{else if product.can_purchase '===' false}}OutOfStock{{else}}InStock{{/if}}",
            "url" : "{{product.url}}",
            "priceValidUntil": "{{moment (moment add='31536000000') 'YYYY-MM-DD' }}"
        }
```

**AFTER:**
```
        "offers": {
            "@type": "Offer",
            "priceCurrency": "{{currency_selector.active_currency_code}}",
            {{#if product.price.price_range}}
            "minPrice": "{{#if product.price.price_range.min.with_tax}}{{product.price.price_range.min.with_tax.value}}{{else}}{{product.price.price_range.min.without_tax.value}}{{/if}}",
            "maxPrice": "{{#if product.price.price_range.max.with_tax}}{{product.price.price_range.max.with_tax.value}}{{else}}{{product.price.price_range.max.without_tax.value}}{{/if}}",
            {{/if}}
            "price": "{{#if product.price.with_tax}}{{product.price.with_tax.value}}{{else}}{{product.price.without_tax.value}}{{/if}}",
            "itemCondition" : "https://schema.org/{{#if product.condition}}{{product.condition}}{{else}}New{{/if}}Condition",
            "availability" : "https://schema.org/{{#if product.pre_order}}PreOrder{{else if product.out_of_stock}}OutOfStock{{else if product.can_purchase '===' false}}OutOfStock{{else}}InStock{{/if}}",
            "url" : "{{product.url}}",
            "priceValidUntil": "{{moment (moment add='31536000000') 'YYYY-MM-DD' }}"
        }
```

**Comment by Tiggerito** [comment]
That's how we do it.

**Comment by Tiggerito** [comment]
An update here:

minPrice and maxPrice are not part of an Offer, but they can be added to a PriceSpecification.

Google does support AggregateOffer, which has a lowPrice and highPrice, but it is not appropriate for BigCommerce stores. It is for product aggregators. Its use may exclude a product from merchant listings. So I would not use AggregateOffer.

https://developers.google.com/search/docs/appearance/structured-data/product#aggregate-offer-properties

**Comment by RobDNZ** [comment]
Hello,

We ran into the same problem and have implemented the same fix including adding the PriceSpecification element as follows:

`"offers": {
            "@type": "Offer",
            "priceCurrency": "{{currency_selector.active_currency_code}}",
            {{#if product.price.price_range}}
            "priceSpecification": {
                "minPrice": "{{#if product.price.price_range.min.with_tax}}{{product.price.price_range.min.with_tax.value}}{{else}}{{product.price.price_range.min.without_tax.value}}{{/if}}",
                "maxPrice": "{{#if product.price.price_range.max.with_tax}}{{product.price.price_range.max.with_tax.value}}{{else}}{{product.price.price_range.max.without_tax.value}}{{/if}}"
            },
            {{/if}}
            "price": "{{#if product.price.with_tax}}{{product.price.with_tax.value}}{{else}}{{product.price.without_tax.value}}{{/if}}",
            "itemCondition" : "https://schema.org/{{#if product.condition}}{{product.condition}}{{else}}New{{/if}}Condition",
            "availability" : "https://schema.org/{{#if product.pre_order}}PreOrder{{else if product.out_of_stock}}OutOfStock{{else if product.can_purchase '===' false}}OutOfStock{{else}}InStock{{/if}}",
            "url" : "{{product.url}}",
            "priceValidUntil": "{{moment (moment add='31536000000') 'YYYY-MM-DD' }}"
        }`

After this the code validates correcly both in the schema.org validator and google rich results test.

https://developers.google.com/search/docs/appearance/structured-data

**Comment by RobDNZ** [comment]
I don't think we can actually test this before it is pushed to the live store but we will definitely test this first and monitor this after deployment to production.

**Comment by RobDNZ** [comment]
The changes have been deployed to our live store and our marketing department has confirmed that they are seeing the expected improvements. In our case the improvements are even bigger because this store did not have the JSON-LD structure yet (they still had the old meta tag stuff).

**Comment by BirddogLighting** [comment]
The product schema hasn't been corrected for product with price range.

**Bug: Product JSON-LD schema uses `Offer` type with `AggregateOffer` properties when a price range is present**

**File:** `templates/components/products/schema.html` (or equivalent schema template)

**Describe the bug**

When a product has variant-driven pricing, `product.price.price_range` is populated and the schema template correctly emits `lowPrice` and `highPrice`. However, the `@type` remains hardcoded as `"Offer"` in both branches of the conditional. `lowPrice` and `highPrice` are properties of `schema.org/AggregateOffer`, not `schema.org/Offer`, making the output invalid structured data.

Google's Rich Results Test and Schema.org validators will flag this as an unrecognized property error on the `Offer` type, which can prevent product rich results from being eligible.

**Steps to reproduce**

1. Create a product with multiple variants that produce a price range (e.g. different SKUs at different prices)
2. View page source and inspect the `application/ld+json` block
3. Observe `@type: "Offer"` alongside `lowPrice` / `highPrice`
4. Paste into https://validator.schema.org — it will report `lowPrice` and `highPrice` as invalid properties for `Offer`

**Expected behavior**

When `product.price.price_range` is present, `@type` should be `"AggregateOffer"`. When it is absent, `@type` should remain `"Offer"`.

**Suggested fix**

```handlebars
"offers": {
    {{#if product.price.price_range}}
    "@type": "AggregateOffer",
    "lowPrice": "{{#if product.price.price_range.min.with_tax}}{{product.price.price_range.min.with_tax.value}}{{else}}{{product.price.price_range.min.without_tax.value}}{{/if}}",
    "highPrice": "{{#if product.price.price_range.max.with_tax}}{{product.price.price_range.max.with_tax.value}}{{else}}{{product.price.price_range.max.without_tax.value}}{{/if}}",
    {{else}}
    "@type": "Offer",
    "price": "{{#if product.price.with_tax}}{{product.price.with_tax.value}}{{else}}{{product.price.without_tax.value}}{{/if}}",
    "priceValidUntil": "{{moment (moment add='31536000000') 'YYYY-MM-DD' }}",
    {{/if}}
    "priceCurrency": "{{currency_selector.active_currency_code}}",
    "itemCondition": "https://schema.org/{{#if product.condition}}{{product.condition}}{{else}}New{{/if}}Condition",
    "availability": "https://schema.org/{{#if product.pre_order}}PreOrder{{else if product.out_of_stock}}OutOfStock{{else if product.can_purchase '===' false}}OutOfStock{{else}}InStock{{/if}}",
    "url": "{{product.url}}"
}
```

Note: `priceValidUntil` is also moved inside the `Offer`-only branch, as it is not a valid property of `AggregateOffer`.

**References**
- https://schema.org/AggregateOffer
- https://schema.org/Offer
- https://developers.google.com/search/docs/appearance/structured-data/product
